### DISCUSSION TITLE
Github Authentication and/or Brew Storage

### DISCUSSION THREAD

**Discussion by dbolack-ab** [naturalcrit/homebrewery#3350] [Ideas]
This is a general feature thread about the desirability of three features for homebrewery.

1.  Authentication with Github: This has two parts. If A is not implemented, B will be necessary to support of items 2 and 3, if implemented.
    A. Add Github as an authentication system as we do with Google Auth
    B. Add a system for storing a Github OAuth token for use with github.
2.  Storage of user Brews in a Github based Repository: Create a fixed name repository and store each brew in the repository. File alteration can be achieved directly via the Github API.
3. Storage of user Brews in a Github Gist: Create user brews as distinct Gists. File alteration can be achieved directly via the Github API.  

There are two main differences between 2 and 3. The first is that a gist is functionally a git repository with all that goes with that. Creating gists is functionally the same as if step 2 created a new repository for each brew, except for the simplified user interface ( which the users might not even touch )

Thoughts?

**Comment by 5e-Cleric** [comment]
Given what you have written, maybe gists is the way to go, a few points i'd like to have in mind:
- In case of conflicts, the newest version should always prevail.
- Do Gists have a public/private setting?
- Are Gists accessible via UI, or would users need HB to find their brews?

**Comment by dbolack-ab** [comment]
> * In case of conflicts, the newest version should always prevail.

I believe that newest always shows as the top and we'd have to go to some effort to not load that version.

> * Do Gists have a public/private setting?

They do. As I read it, you cannot unpublic a gist but you can move one from private to public. My assumption was we would leave them private and load them via server API call. If they were public we could load them in the browser though so, I dunno. If we needed the functionality to unpublic one, we'd need to create a new gist and recreate the history, but as a private gist. 

> * Are Gists accessible via UI, or would users need HB to find their brews?

Gists are available via the Github UI. They show up as an item on the user menu. They can even use that to peruse previous revisions.

**Comment by dbolack-ab** [comment]
Bonus feature I discovered today. If you name the file you push into the gist, it shows that on the gist listings instead of the gistid. By naming the file the edited I made it very easy to find on the gist listing if I needed to.
